transport/src/main/java/io/netty/channel/AutoFlushHandler.java

```diff
@@ -0,0 +1,177 @@
+ * Flushing batches is arbitrary and has to be deviced for each stream in an arbitrary way.  
+ *
+ * This handler addresses the above usecase by providing a way to batch writes and flush them at the following points:
+ * <ol>
+ * <li>At the end of the eventloop iteration.</li>
+ * <li>If the channel becomes unwritable determined via {@link #channelWritabilityChanged(ChannelHandlerContext)}
+ * callback.</li>
+ * </ol>
+ *
+ * Any explicit flushes are honored as usual.
+ *
+ * <h2>Restrictions</h2>
+ *
+ * Due to limitation of existing APIs, {@link AutoFlushHandler} can only be used if the following conditions are met:
+ *
+ * <ol>
+ *     <li>Channel is registered with an {@link EventLoop} that extends {@link SingleThreadEventLoop}.</li>
+ *     <li>{@link ChannelPipeline} is an instance of {@link DefaultChannelPipeline}</li>
+ *     <li>This handler is not executed by any executor apart from the channel's eventloop.</li>
+ * </ol>
+ *
+ * If any of the above conditions are not met, this handler will fail on
+ * {@link #channelRegistered(ChannelHandlerContext)}.
+ *
+ * <h2>Recommendations</h2>
+ *
+ * The {@link AutoFlushHandler} should be put as first {@link ChannelHandler} in the
+ * {@link ChannelPipeline} to have the best effect.
+ */
+public class AutoFlushHandler extends ChannelDuplexHandler implements Runnable {
+    private boolean flushPending;
+    private boolean flushScheduled;
+    private SingleThreadEventLoop eventLoop;
+    private Channel channel;
+    private DefaultChannelPipeline pipeline;
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        channel = ctx.channel();
+        ChannelPipeline pipeline = ctx.channel().pipeline();
+        if (!(pipeline instanceof DefaultChannelPipeline)) {
+            throw new IllegalStateException(
+                    "Auto flush is only supported with " + DefaultChannelPipeline.class.getName());
+        }
+        this.pipeline = (DefaultChannelPipeline) pipeline;
+        EventLoop eventLoop = ctx.channel().eventLoop();
+        if (!(eventLoop instanceof SingleThreadEventLoop)) {
+            throw new IllegalStateException(
+                    "Auto flush is only supported for " + SingleThreadEventLoop.class.getName() + " eventloops.");
+        }
+        this.eventLoop = (SingleThreadEventLoop) eventLoop;
+        if (this.eventLoop != ctx.executor()) {
+            throw new IllegalStateException(AutoFlushHandler.class.getName() + " must run on the channel eventloop.");
+        }
+        super.channelRegistered(ctx);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        flushPending = true;
+        super.write(ctx, msg, promise);
+        scheduleFlushIfRequired();
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) throws Exception {
+        flushIfPending(ctx);
```

Comment:
I gave my reasons in the other PR as to why not, so let's just agree to disagree.